### DISCUSSION TITLE
AOC 2022 Day Two

### DISCUSSION THREAD

**Discussion by bogdaaamn** [Virtual-Coffee/virtualcoffee.io#756] [Advent of Code 2022]
Discussion of solutions for [Advent of Code Day 2](https://adventofcode.com/2022/day/2) here

---

### VC Advent of code leaderboard
Go to [leaderboard/private](https://adventofcode.com/2022/leaderboard/private) and use join code `363496-e6f42283`

**Comment by bogdaaamn** [comment]
Mine is super super silly https://github.com/BogDAAAMN/advent-of-code-2022/blob/main/day-02.js

But I can't wait to see how you folks did it better and more efficient 😂

**Reply by turingtestable** [reply]
I am still not sure if using a hash/dict is better or just having the control structure that actually makes sense is better, so i wouldn't consider your solution silly.

**Reply by bogdaaamn** [reply]
yeah I thought about that for a second, but not sure if it would improve readability or make it works 😭

**Reply by turingtestable** [reply]
I actually thought about doing two hashes one for my strategy `{X:1, Y:2, Z:3}` and one for win/loss/draw so it was easier to understand, but ultimately it felt like a thing i would just have to refactor in real life,  so i switched back to just one hash.

**Reply by bogdaaamn** [reply]
OHHHHH that would make a lot of sense tbh, I would like to see that

**Reply by danieltott** [reply]
I sort of did both - 
```ts
type PlayerAValue = "A" | "B" | "C";
type PlayerBValue = "X" | "Y" | "Z";

enum Actions {
  Rock = 1,
  Paper = 2,
  Scissors = 3,
}

const actionDict = {
  A: Actions.Rock,
  B: Actions.Paper,
  C: Actions.Scissors,
  X: Actions.Rock,
  Y: Actions.Paper,
  Z: Actions.Scissors,
};
```

And then below there's some control structures

**Reply by tkshill** [reply]
Comments so you don't lose track seems super sensible

**Comment by turingtestable** [comment]
Python [Part1](https://github.com/turingtestable/AdventOfCode2022/blob/master/challenge2/solution.py) [Part2](https://github.com/turingtestable/AdventOfCode2022/blob/master/challenge2/solution2.py) 

Raise your hand if you had to write down the values of everything on a piece of paper to make the hash of values.

**Reply by doleraj** [reply]
Yyyyyyyup

**Reply by bogdaaamn** [reply]
✋🏼

**Reply by MCHadley** [reply]
YUP

**Comment by turingtestable** [comment]
Here's my [ruby solution](https://github.com/turingtestable/AdventOfCode2022/blob/master/challenge2/solution.rb) realized reducing would have been better in python but not going to change my solution now after posting.

**Reply by bogdaaamn** [reply]
I love the dict solution!!

**Reply by tkshill** [reply]
This is so smart! Such an "understood the assignment" solution

**Reply by danieltott** [reply]
Damn that's a really nice solution

**Reply by davidalpert** [reply]
did you figure out all those values for each combination by hand?

**Reply by turingtestable** [reply]
Yep, started with pencil and paper.

**Comment by doleraj** [comment]
Holy mother of fuck I spent WAY TOO LONG operating under the assumption that A/B/C was _your_ play, not your opponent's play. The worst part? The example given _works_ if you make that assumption.

Anyhow. https://github.com/doleraj/adventofcode/blob/main/aoc2022/src/day02/index.ts

**Reply by turingtestable** [reply]
Oh man i have spent way too long on that kind of thing before, they are worded clearly in retrospect, but in the moment it is easy to gloss over the one critical piece of information. Rough that the example works either way.

**Reply by bogdaaamn** [reply]
Totally forgot switches are a thing 😂 Looks cleaner than my nested if elses

**Reply by doleraj** [reply]
I'm realizing as I look at it that I should have used .map over the lines instead of the index nonsense. I was poking at previous years where I'd used Kotlin, where `for (foo in bar)` gets you the items in the list, instead of JS/TS where it just gets you... the index. For some bizarre reason

**Reply by tkshill** [reply]
Love the conciseness here! But yeah, this was totally a problem where I lost a ton of time because I just... read it wrong.

**Comment by danieltott** [comment]
Here's my solution! https://github.com/danieltott/aoc2022/blob/main/src/day02/index.ts

Had some fun with `enum`s - it was cool being able to refer to `Actions.Rock` etc.

I also used nested switches. I really feel like there's a neeater way to do this...

**Reply by bogdaaamn** [reply]
oh wow you went all in typescript!!! I love the solution very neat and readable

**Reply by doleraj** [reply]
There's not a super great way to do this, because TS enums SUCK compared to other languages. It'd be so much easier if you could put a compare function on the enum itself to get the score.

**Reply by bogdaaamn** [reply]
sorry not sorry javascript

**Reply by tkshill** [reply]
I like this solution. Feels like a nice mix of elegant and practical. Agreed that enums in js/ts are weird which is why I prefer unions.

**Reply by danieltott** [reply]
Yeah - I could have avoided the `actionDict` if I could compare enum values - but I guess I get why that is? enums are still pretty new to me

**Comment by tkshill** [comment]
I'm not sure how I feel about this answer. There's some stuff I'm glad I did, but I spent wayyyyy too long updating my scaffolding for this whole thing, which had nothing to do with today at all.

More F#, fp style goodness. 

[answer](https://github.com/tkshill/AdventOfCode2022/blob/main/AOC2022/Day02.fs)

I also have a [utility file](https://github.com/tkshill/AdventOfCode2022/blob/main/AOC2022/Accessories.fs) with some convenient glue funcs I keep using throughout these things.

**Reply by davidalpert** [reply]
I'm going to have to study with you, @tkshill. that is much more elegant than my F#

**Comment by MCHadley** [comment]
It works, but it's not pretty? I did it in 2 parts rather than trying to refactor my code to cover both parts. And I did it in JS

[answer](https://github.com/MCHadley/adventofcode/tree/master/2022/day2)

**Reply by tkshill** [reply]
I can't open this link? Not sure if it's me or if the repo is private

**Comment by clandau** [comment]
JS https://github.com/clandau/advent-of-code/blob/main/2022/day-2/2.js

**Reply by tkshill** [reply]
This is great! I really like the use of modulo and the offset logic. I thought about it for a bit but I couldn't figure out how to get it working for my answer. Concise JS is always fun to look at 😄

**Reply by danieltott** [reply]
Ugh yes this is doooope

**Reply by bogdaaamn** [reply]
OF COURSE THERE IS A MODULO SOLTUION

**Reply by bogdaaamn** [reply]
I love everything about this, the modulo, the offsets. this is really what I pictured being a good js solution is. just never got the brain power to do it lol

**Comment by davidalpert** [comment]
here's another F# take at it
- domain: https://github.com/davidalpert/advent-of-code-da/blob/2022/2022/day02_rock_paper_scissors.fs 
- calculations: https://github.com/davidalpert/advent-of-code-da/blob/2022/2022/day02_tests.fs#L2522-L2554

**Reply by tkshill** [reply]
I aspire to have your robustness. My solutions takes some shortcuts where yours goes for completeness. I think your typing win/loss/draw explicitly adds more readability. And as always you remind me that I should put more effort in my code infrastructure. I'm trying to use dotnet in a more full featured way so I might ask you some questions about your project configurations

**Comment by Cerchie** [comment]
It's so inspiring to see all these creative solutions! I'll post mine once I make my repo public again.

**Reply by Cerchie** [reply]
Here's mine fyi: https://github.com/Cerchie/adventofcode2022/blob/main/day-02/index.js

**Reply by Cerchie** [reply]
12 'else if's 🤣
but it worked and was easy to change up for round 2

**Reply by bogdaaamn** [reply]
UH OH welcome to the if else club 😂😂

**Comment by sadiejay** [comment]
Here's mine! I got part 1 but the change up for part 2 felt targeted haha Gonna try day 3 and come back to this, but would appreciate any new strategy hah 

https://github.com/sadiejay/aoc2022/tree/main/dec2

**Reply by tkshill** [reply]
Looks good! Yeah, I iterate in these depending on my whims

**Reply by sadiejay** [reply]
@tkshill Thank you taking a look!! Andddd I just part 2!!!! 💃🏾

Now going to day 3 for real lol
